shapely/ops.py

```diff
@@ -684,3 +686,155 @@ def orient(geom, sign=1.0):
+    ...:         LinearRing(([0, 0], [1, 1], [2, 2], [0,0])),
+    ...:     ],
+    ...:     return_index=True,
+    ...: )
+    >>> segments.tolist(), index.tolist()
+    ([<LINESTRING (0 0, 1 1)>,
+      <LINESTRING (1 1, 2 2)>,
+      <LINESTRING (0 0, 1 1)>,
+      <LINESTRING (1 1, 2 2)>,
+      <LINESTRING (2 2, 0 0)>],
+     [0, 0, 1, 1, 1])
+
+    By default the third dimension (Z) is ignored.
+
+    >>> get_segments(LineString(([0, 0, 1], [1, 1, 1], [2, 2, 1]))).tolist()
+    [<LINESTRING (0 0, 1 1)>, <LINESTRING (1 1, 2 2)>]
+    >>> get_segments(
+    ...     LineString(([0, 0, 1], [1, 1, 1], [2, 2, 1])), include_z=True,
+    ... ).tolist()
+    [<LINESTRING Z (0 0 1, 1 1 1)>, <LINESTRING Z (1 1 1, 2 2 1)>]
+
+
+    If geometries don't have a Z dimension, these values will be NaN.
+
+    >>> get_segments(LineString(([0, 0], [1, 1], [2, 2])), include_z=True).tolist()
+    [<LINESTRING Z (0 0 NaN, 1 1 NaN)>, <LINESTRING Z (1 1 NaN, 2 2 NaN)>]
+
+    """
+
+    def _validate_input_feature(g):
+        """Ensure input feature is linear."""
+        if g.geom_type not in valid_geometry_types:
+            raise GeometryTypeError(
+                f"Check input. Getting segments from '{g.geom_type}' is not supported."
+            )
+
+    def _validate_input(g):
+        """Ensure all input features are linear."""
+        if hasattr(g, "geom_type"):
+            _validate_input_feature(g)
+            single_geom = True
+        else:
+            np.vectorize(_validate_input_feature)(g)
+            single_geom = False
+
+        return single_geom
+
+    def _determine_signature(coords_by_feature):
+        """Determine if a signature is needed for the vectorization."""
+        array_signature = "(n,m),()->(k)"
+        signature = array_signature if len(coords_by_feature) == 1 else None
+        if not signature:
+            try:
+                np.array(coords_by_feature)
+                signature = array_signature
+            except ValueError:
+                pass
+
+        return signature
+
+    def _isolate_segments(coords, include_z, **kwargs):
+        """Isolate segments from individual input linear features."""
+        coords_stacked = np.column_stack((coords[:-1], coords[1:]))
+
+        n_segments = coords_stacked.shape[0]
+        n_dims = 2
+        if include_z:
+            n_dims += 1
+
+        return linestrings(coords_stacked.reshape(n_segments, 2, n_dims), **kwargs)
+
+    def _vectorized_isolate_segments(coords, idx, include_z, **kwargs):
+        """Isolate segments from all input linear features."""
+        n_xys = coords.shape[0]
+        idx_splitter = np.unique(idx.reshape(n_xys, 1), return_index=True)[1][1:]
+        xys_by_input_feature = np.split(xys, idx_splitter)
+        signature = _determine_signature(xys_by_input_feature)
+
+        return np.concatenate(
+            np.vectorize(_isolate_segments, signature=signature, otypes=[np.ndarray])(
```

Comment:
It will probably be hard, but try to do this without the `np.vectorize`. We can discuss about it if you can't find a solution.